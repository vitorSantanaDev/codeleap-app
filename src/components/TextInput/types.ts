import { InputHTMLAttributes } from 'react'

export type TextInputProps = {
  label?: string
  disabled?: boolean
  initialValue?: string
  handleChange?: (value: string) => void
} & InputHTMLAttributes<HTMLInputElement>
