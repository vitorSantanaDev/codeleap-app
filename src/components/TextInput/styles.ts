import styled, { css } from 'styled-components'

export const Wrapper = styled.div``

export const InputElement = styled.input`
  ${({ theme }) => css`
    width: 100%;
    display: block;
    outline: none;
    color: ${theme.colors.black};
    padding: ${theme.spacings.xxsmall};
    border-radius: ${theme.border.radius};
    border: 1px solid ${theme.colors.dark_gray};

    &::placeholder {
      color: ${theme.colors.light_gray};
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    display: inline-block;
    color: ${theme.colors.black};
    font-size: ${theme.font.sizes.medium};
    margin-bottom: ${theme.spacings.xxsmall};
  `}
`
