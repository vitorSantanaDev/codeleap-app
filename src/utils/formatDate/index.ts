import {
  format,
  parseISO,
  minutesToHours,
  secondsToMinutes,
  differenceInSeconds
} from 'date-fns'

export const formattingStringWithTheTimePassedBetweenTwoDate = (
  hours: number,
  minutes: number,
  seconds: number
): string => {
  if (hours < 1 && minutes < 1) {
    return `${seconds} seconds ago`
  }

  let formattedTime = ''

  if (hours > 0) {
    formattedTime = `${hours} ${hours > 1 ? 'hours' : 'hour'} ago`
  } else if (hours < 1) {
    formattedTime = `${minutes} minutes ago`
  }

  return formattedTime
}

export const rangeOfTimePassedBetweenTwoDates = (
  start: string,
  end: string
): string => {
  const startDate = parseISO(start)
  const endDate = parseISO(end)

  const allSeconds = differenceInSeconds(endDate, startDate)
  const allMinutes = secondsToMinutes(allSeconds)

  const minutesInSeconds = allMinutes * 60
  const seconds = allSeconds - minutesInSeconds

  const hours = minutesToHours(allMinutes)

  const hoursInMinutes = hours * 60

  const minutes = allMinutes - hoursInMinutes

  return hours > 24
    ? format(new Date(startDate), 'dd/MM/yyyy')
    : formattingStringWithTheTimePassedBetweenTwoDate(hours, minutes, seconds)
}
