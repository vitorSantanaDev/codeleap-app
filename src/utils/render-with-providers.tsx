import { Provider } from 'react-redux'
import React, { PropsWithChildren } from 'react'

import theme from 'styles/theme'
import { ThemeProvider } from 'styled-components'
import { RenderOptions, render } from '@testing-library/react'

import { RootState } from 'redux/store'
import { userReducer } from 'redux/actions/user-slice'
import { PreloadedState, configureStore } from '@reduxjs/toolkit'
import { QueryClient, QueryClientProvider } from 'react-query'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  store?: any
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = configureStore({
      reducer: { user: userReducer },
      preloadedState: initialState
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <QueryClientProvider client={new QueryClient()}>
        <Provider store={store}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
      </QueryClientProvider>
    )
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
