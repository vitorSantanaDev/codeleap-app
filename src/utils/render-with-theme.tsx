import React, { ReactElement } from 'react'

import theme from 'styles/theme'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

export const renderWithTheme = (component: ReactElement) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
