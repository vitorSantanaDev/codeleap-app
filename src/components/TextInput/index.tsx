import React, { useState } from 'react'
import { TextInputProps } from './types'

import * as S from './styles'

const TextInput: React.FC<TextInputProps> = ({
  name,
  label,
  handleChange,
  initialValue = '',
  ...restProps
}) => {
  const [value, setValue] = useState(initialValue)

  const onChangeInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.currentTarget.value
    setValue(newValue)

    !!handleChange && handleChange(newValue)
  }

  return (
    <S.Wrapper>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputElement
        type="text"
        name={name}
        value={value}
        onChange={onChangeInput}
        {...(label ? { id: name } : {})}
        {...restProps}
      />
    </S.Wrapper>
  )
}

export default TextInput
