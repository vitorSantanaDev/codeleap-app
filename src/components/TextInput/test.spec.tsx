import React from 'react'
import { renderWithTheme } from 'utils/render-with-theme'
import TextInput from '.'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

describe('<TextInput />', () => {
  it('should render input with label', () => {
    renderWithTheme(<TextInput label="Title" name="title" />)

    const input = screen.getByLabelText(/title/i)
    expect(input).toBeInTheDocument()
  })

  it('should render input without label', () => {
    renderWithTheme(<TextInput name="title" />)

    const input = screen.queryByLabelText(/title/i)
    expect(input).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextInput placeholder="Title" />)

    const input = screen.getByPlaceholderText('Title')
    expect(input).toBeInTheDocument()
  })

  it('should there must ne changes in the value whwn typing in the input', async () => {
    const handleChange = jest.fn()
    renderWithTheme(
      <TextInput handleChange={handleChange} label="Title" name="title" />
    )

    const input = screen.getByRole('textbox')
    const newValue = 'My title'

    userEvent.type(input, newValue)

    await waitFor(() => {
      expect(input).toHaveValue(newValue)
      expect(handleChange).toHaveBeenCalledTimes(newValue.length)
    })

    expect(handleChange).toHaveBeenCalledWith(newValue)
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(
      <TextInput label="Title" name="title" />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
