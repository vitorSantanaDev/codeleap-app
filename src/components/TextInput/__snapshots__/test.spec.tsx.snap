// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should match snapshot 1`] = `
.c1 {
  width: 100%;
  display: block;
  outline: none;
  color: #000000;
  padding: 0.8rem;
  border-radius: 0.8rem;
  border: 1px solid #777777;
}

.c1::-webkit-input-placeholder {
  color: #CCCCCC;
}

.c1::-moz-placeholder {
  color: #CCCCCC;
}

.c1:-ms-input-placeholder {
  color: #CCCCCC;
}

.c1::placeholder {
  color: #CCCCCC;
}

.c0 {
  display: inline-block;
  color: #000000;
  font-size: 1.6rem;
  margin-bottom: 0.8rem;
}

<div
  class=""
>
  <label
    class="c0"
    for="title"
  >
    Title
  </label>
  <input
    class="c1"
    id="title"
    name="title"
    type="text"
    value=""
  />
</div>
`;
